store/store.go

```diff
@@ -1255,9 +1257,69 @@ func (s *Store) downloadDelta(name string, downloadDir string, downloadInfo *sna
 		return "", err
 	}
 
+	// TODO: Check sha3_384
 	return deltaPath, nil
 }
 
+// applyDelta generates a target snap from the current snap and a downloaded delta.
+var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo) (string, error) {
+	currentSnapName := fmt.Sprintf("%s_%d.snap", name, deltaInfo.FromRevision)
+	currentSnapPath := filepath.Join(dirs.SnapBlobDir, currentSnapName)
+
+	if !osutil.FileExists(currentSnapPath) {
+		return "", fmt.Errorf("the source snap at revision %d could not be found at %s", deltaInfo.FromRevision, currentSnapPath)
+	}
+
+	if deltaInfo.Format != "xdelta" {
+		return "", fmt.Errorf("unsupported delta format %q. Currently only \"xdelta\" format is supported", deltaInfo.Format)
+	}
+
+	targetSnapName := fmt.Sprintf("%s_%d_via_delta_%d_%d.snap", name, deltaInfo.ToRevision, deltaInfo.FromRevision, deltaInfo.ToRevision)
+
+	// Create a temporary file only to get the unique path.
+	tmpfile, err := ioutil.TempFile("", targetSnapName)
+	if err != nil {
+		return "", err
+	}
+	targetSnapPath := tmpfile.Name()
+	if err = tmpfile.Close(); err != nil {
+		return "", err
+	}
+
+	xdeltaArgs := []string{"patch", deltaPath, currentSnapPath, targetSnapPath}
+	cmd := exec.Command("xdelta", xdeltaArgs...)
+
+	if err = cmd.Run(); err != nil {
+		return "", err
+	}
+
+	return targetSnapPath, nil
+}
+
+// downloadAndApplyDelta downloads and then applies the delta to the current snap.
+func (s *Store) downloadAndApplyDelta(name string, downloadInfo *snap.DownloadInfo, pbar progress.Meter, user *auth.UserState) (path string, err error) {
+	deltaInfo := &downloadInfo.Deltas[0]
+	workingDir, err := ioutil.TempDir("", name+"-deltas")
+	if err != nil {
+		return "", err
+	}
+	defer os.RemoveAll(workingDir)
+
+	deltaPath, err := s.downloadDelta(name, workingDir, downloadInfo, pbar, user)
+	if err != nil {
+		return "", err
+	}
+
+	logger.Debugf("Successfully downloaded delta for %s at %s", name, deltaPath)
```

Comment:
Done those two, and I don't see any other interpolations of name in the diff (well, except for constructing file names).